In [10]:
import pandas as pd
import seaborn as sns

## Reading files

In [11]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [12]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [13]:
wh = pd.concat([wh2018,wh2019])

In [14]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [15]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

Mostrar la cantidad de ulos que hay 

In [ ]:
print(wh.isnull().sum())

In [ ]:
wh.to_csv('2018_v2.csv',index=False)

In [18]:
print(wh2019.isnull().sum())

Overall rank                     0
Country or region                0
Score                            2
GDP per capita                   5
Social support                  11
Healthy life expectancy         10
Freedom to make life choices     2
Generosity                       7
Perceptions of corruption        8
year                             0
dtype: int64


In [ ]:
wh.isnull().any(axis=0).sum()

In [21]:
wh2018.isnull().any(axis=0)

Overall rank                    False
Country or region               False
Score                            True
GDP per capita                   True
Social support                   True
Healthy life expectancy          True
Freedom to make life choices     True
Generosity                       True
Perceptions of corruption        True
year                            False
dtype: bool

In [22]:
wh2019.isnull().any(axis=0)

Overall rank                    False
Country or region               False
Score                            True
GDP per capita                   True
Social support                   True
Healthy life expectancy          True
Freedom to make life choices     True
Generosity                       True
Perceptions of corruption        True
year                            False
dtype: bool

El total de nulos que hay 

In [27]:
wh.isnull().any(axis=0).sum()

7

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

La descripcion de los dos años

In [24]:
resumen= wh.groupby('year')

In [28]:
resumen.describe()

Overall rank                                                    Score  \
            count  mean        std  min    25%   50%     75%    max  count   
year                                                                         
2018        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  151.0   
2019        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  154.0   

               ... Generosity        Perceptions of corruption            \
         mean  ...        75%    max                     count      mean   
year           ...                                                         
2018  5.35406  ...      0.234  0.598                     149.0  0.112859   
2019  5.39624  ...      0.252  0.566                     148.0  0.111189   

                                                     
           std  min     25%     50%      75%    max  
year                                                 
2018  0.098108  0.0  0.0500  0.0820  0.13800  0.457  
2019  0.094986  0.0  0.0465  0.0865  0.14125  0.453  

[2 rows x 64 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [41]:
wh["NaN"].mean()





KeyError: 'NaN'

In [43]:
wh["NaN"].replace(np.nan,0)

KeyError: 'NaN'

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro